tree-sitter/src/TreeSitter/GenerateSyntax.hs

```diff
@@ -82,14 +92,13 @@ syntaxDatatype language datatype = skipDefined $ do
 
 
 -- | Create TH-generated SymbolMatching instances for sums, products, leaves
-symbolMatchingInstance :: Ptr TS.Language -> Name -> String -> Q [Dec]
-symbolMatchingInstance language name str = do
-  tsSymbol <- runIO $ withCString str (TS.ts_language_symbol_for_name language)
-  tsSymbolType <- toEnum <$> runIO (TS.ts_language_symbol_type language tsSymbol)
+symbolMatchingInstance :: [(String, Named)] -> Name -> Named -> String -> Q [Dec]
+symbolMatchingInstance allSymbols name named str = do
+  let tsSymbols = elemIndices (str, named) allSymbols
+  let names = intercalate ", " $ fmap (fst . (!!) allSymbols) tsSymbols
   [d|instance TS.SymbolMatching $(conT name) where
-      showFailure _ node = "Expected " <> $(litE (stringL (show name))) <> " but got " <> show (TS.fromTSSymbol (nodeSymbol node) :: $(conT (mkName "Grammar.Grammar")))
-      symbolMatch _ node = TS.fromTSSymbol (nodeSymbol node) == $(conE (mkName $ "Grammar." <> TS.symbolToName tsSymbolType str))|]
-
+      showFailure _ node = "expected " <> $(litE (stringL (show names))) <> " but got " <> show (TS.fromTSSymbol (nodeSymbol node) :: $(conT (mkName "Grammar.Grammar")))
```

Comment:
Currently, this will have a different naming convention for the `expected` than for the `got`, right? Like `Expected yield_expression but got AwaitExpression`. Or is that fixed? I think ideally, we'd drop the auto-generated enum entirely, and say `Expected yield_expression but got await_expression`.

Either way, I don't think it needs to block this PR necessarily.

Reply:
Yeah we could make this better, but let's do it in another PR. I got a bit confused with the compile time and template haskell vs. runtime stuff here to get the error messages right.